refactor: mining interface 30955 followups

Sjors opened the pull request:
This PR implements the refactors suggested in https://github.com/bitcoin/bitcoin/pull/30955#pullrequestreview-2354931253.

Sjors committed:
Rename merkle branch to path

Sjors committed:
Drop TransactionMerklePath default position arg

Sjors committed:
refactor: use CTransactionRef in submitSolution

ryanofsky reviewed (approved):
Code review ACK 058862581085316927287817b2af01e8f4765a1d. Looks good, thanks for the cleanups!

TheCharlatan commented on src/consensus/merkle.cpp:
```diff
@@ -84,8 +84,8 @@ uint256 BlockWitnessMerkleRoot(const CBlock& block, bool* mutated)
 /* This implements a constant-space merkle root/path calculator, limited to 2^32 leaves. */
-static void MerkleComputation(const std::vector<uint256>& leaves, uint256* proot, bool* pmutated, uint32_t branchpos, std::vector<uint256>* pbranch) {
-    if (pbranch) pbranch->clear();
+static void MerkleComputation(const std::vector<uint256>& leaves, uint256* proot, bool* pmutated, uint32_t leaf_pos, std::vector<uint256>* path) {
```
Nit: Since this is a style fix in the first place already, could also put the bracket on a new line to make clang-format happy.

Sjors replied:
Will do if I need to retouch.

Sjors replied:
Done

sedited reviewed (approved):
ACK 058862581085316927287817b2af01e8f4765a1d

itornaza reviewed (approved):
ACK 058862581085316927287817b2af01e8f4765a1d

Clean refactor in Merkle context variable descriptions from using `branch` to `path`. I also find not using default parameters more beneficial as in the current case of the `TransactionMerklePath`'s, `position` argument. 

Also, did a clean built locally and run all unit, functional and extended tests which all pass.

tdb3 reviewed (approved):
code review ACK 058862581085316927287817b2af01e8f4765a1d

The updates make the interface clearer. Left one non-blocking comment.

Lightly tested by cherry-picking bitcoin-mine from #30437 to exercise some of the interface (exercising `TransactionMerklePath()`, `ComputeMerklePath()`, and `MerkleComputation()`) . Saw the merkle path being built up as transactions were added.

<details>

<summary>Output from bitcoin-mine</summary>

```
git cherry-pick 1a4ca89fad87464023430ebe02685fa23ed2bb1d
git cherry-pick 8fab41611a473635ca5fc4804d765cc4de25857c
git cherry-pick 99f2f335b6a8c2b5fe518a0c0056eb2a42c0496b
```

```diff
diff --git a/src/bitcoin-mine.cpp b/src/bitcoin-mine.cpp
index 4d5faea23e..473fcaf1b9 100644
--- a/src/bitcoin-mine.cpp
+++ b/src/bitcoin-mine.cpp
@@ -120,5 +120,14 @@ MAIN_FUNCTION
         tfm::format(std::cout, "Tip hash is null.\n");
     }
 
+    std::vector<unsigned char> v{0x00, 0x14, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99, 0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99, 0x00};
+    CScript coinbase_script = CScript() << std::vector<unsigned char>(v.begin(), v.end());
+    std::unique_ptr<interfaces::BlockTemplate> new_block{mining->createNewBlock(coinbase_script)};
+    auto merkle_path{new_block->getCoinbaseMerklePath()};
+    tfm::format(std::cout, "Printing merkle path:\n");
+    for (auto const& b : merkle_path) {
+        tfm::format(std::cout, "%s\n", b.GetHex());
+    }
+
     return EXIT_SUCCESS;
 }
```

```
make -C depends NO_QT=1 MULTIPROCESS=1
HOST_PLATFORM="x86_64-pc-linux-gnu"
cmake -B build --toolchain=depends/$HOST_PLATFORM/toolchain.cmake
cmake --build build -j12
build/src/bitcoin-node -regtest -printtoconsole -debug=ipc -ipcbind=unix
```

```
$ build/src/bitcoin-cli -regtest createwallet test
$ build/src/bitcoin-cli -regtest -generate 101
```



```
$ build/src/bitcoin-mine -regtest
Connected to bitcoin-node
Tip hash is 12dda23b94be0cffe84c094fa30c25e08ead0b3ca2e8b8b738a87d4decc96d04.
Printing merkle path:

$ build/src/bitcoin-cli -named sendtoaddress address=bcrt1qctncrkr9ucfvhxkkvf70zegrkqm0k8jwxg7m9u amount=1.2 fee_rate=2
354f5f8a7b741395656417958d3339b8affb74f2bc7c93efc8c2710207c74e79

$ build/src/bitcoin-mine -regtest
Connected to bitcoin-node
Tip hash is 12dda23b94be0cffe84c094fa30c25e08ead0b3ca2e8b8b738a87d4decc96d04.
Printing merkle path:
354f5f8a7b741395656417958d3339b8affb74f2bc7c93efc8c2710207c74e79

$ build/src/bitcoin-cli -named sendtoaddress address=bcrt1qctncrkr9ucfvhxkkvf70zegrkqm0k8jwxg7m9u amount=1.2 fee_rate=2
62b8b4e735316df78125d4119c0ab867c71fe588668bdc7323c0b7707e3235dc

$ build/src/bitcoin-mine -regtest
Connected to bitcoin-node
Tip hash is 12dda23b94be0cffe84c094fa30c25e08ead0b3ca2e8b8b738a87d4decc96d04.
Printing merkle path:
354f5f8a7b741395656417958d3339b8affb74f2bc7c93efc8c2710207c74e79
84cabb7420050b1c3b18a162dccf7526b055b195ac873a0baea3c0c52a2c7e56

$ build/src/bitcoin-cli -named sendtoaddress address=bcrt1qctncrkr9ucfvhxkkvf70zegrkqm0k8jwxg7m9u amount=1.2 fee_rate=2
425686bfda45ef369c7acb2636c5c234084c8361f54ffd39dda3c1a330bd3a22

$ build/src/bitcoin-mine -regtest
Connected to bitcoin-node
Tip hash is 12dda23b94be0cffe84c094fa30c25e08ead0b3ca2e8b8b738a87d4decc96d04.
Printing merkle path:
354f5f8a7b741395656417958d3339b8affb74f2bc7c93efc8c2710207c74e79
8f0b568257b75176c56703ae57036cba1bf2cc23a5909f60d6bc9544473a553f

$ build/src/bitcoin-cli -named sendtoaddress address=bcrt1qctncrkr9ucfvhxkkvf70zegrkqm0k8jwxg7m9u amount=1.2 fee_rate=2
e152a9ac29651ce96e5ee59b648c6e14ff850df1799dfd5453f96e681f3d0267

$ build/src/bitcoin-mine -regtest
Connected to bitcoin-node
Tip hash is 12dda23b94be0cffe84c094fa30c25e08ead0b3ca2e8b8b738a87d4decc96d04.
Printing merkle path:
354f5f8a7b741395656417958d3339b8affb74f2bc7c93efc8c2710207c74e79
8f0b568257b75176c56703ae57036cba1bf2cc23a5909f60d6bc9544473a553f
73a8292c417ccdd779e725e54c71968a7feacaa00f083fc8b08865f0de8caf5a

$ build/src/bitcoin-cli -named sendtoaddress address=bcrt1qctncrkr9ucfvhxkkvf70zegrkqm0k8jwxg7m9u amount=1.2 fee_rate=2
251ff17e509ca64eb4a026b571943de86202420d15ceccfc273bb9806accc6a1

$ build/src/bitcoin-mine -regtest
Connected to bitcoin-node
Tip hash is 12dda23b94be0cffe84c094fa30c25e08ead0b3ca2e8b8b738a87d4decc96d04.
Printing merkle path:
354f5f8a7b741395656417958d3339b8affb74f2bc7c93efc8c2710207c74e79
8f0b568257b75176c56703ae57036cba1bf2cc23a5909f60d6bc9544473a553f
6db4d6f4ed256dff78d636ea2950e801ebc0b8aa9c781374dbf545cc38f5f705

$ build/src/bitcoin-cli -named sendtoaddress address=bcrt1qctncrkr9ucfvhxkkvf70zegrkqm0k8jwxg7m9u amount=1.2 fee_rate=2
65fc3f53395ff343f684c666027ca565a41dda1e8aa69ff06d38baf56e160e00

$ build/src/bitcoin-mine -regtest
Connected to bitcoin-node
Tip hash is 12dda23b94be0cffe84c094fa30c25e08ead0b3ca2e8b8b738a87d4decc96d04.
Printing merkle path:
354f5f8a7b741395656417958d3339b8affb74f2bc7c93efc8c2710207c74e79
8f0b568257b75176c56703ae57036cba1bf2cc23a5909f60d6bc9544473a553f
7d315aaad342ffb8b0c9431fc7601dfd79fa72ea64492f1b27288e83e43ff22e

$ build/src/bitcoin-cli -named sendtoaddress address=bcrt1qctncrkr9ucfvhxkkvf70zegrkqm0k8jwxg7m9u amount=1.2 fee_rate=2
51c21fab60d667d346a2ea81c813efbad181daffe2381c1ab40b3e8541678a0f

$ build/src/bitcoin-mine -regtest
Connected to bitcoin-node
Tip hash is 12dda23b94be0cffe84c094fa30c25e08ead0b3ca2e8b8b738a87d4decc96d04.
Printing merkle path:
354f5f8a7b741395656417958d3339b8affb74f2bc7c93efc8c2710207c74e79
8f0b568257b75176c56703ae57036cba1bf2cc23a5909f60d6bc9544473a553f
e4e50b35186ef0bec23c7df9473856a673dc2cf401c9b6552e540618fecc6954

$ build/src/bitcoin-cli -named sendtoaddress address=bcrt1qctncrkr9ucfvhxkkvf70zegrkqm0k8jwxg7m9u amount=1.2 fee_rate=2
aad26e5184382623d900c06de7ef5cf4eb861b4a40d418a7f80e0f007fa0cb9a

$ build/src/bitcoin-mine -regtest
Connected to bitcoin-node
Tip hash is 12dda23b94be0cffe84c094fa30c25e08ead0b3ca2e8b8b738a87d4decc96d04.
Printing merkle path:
354f5f8a7b741395656417958d3339b8affb74f2bc7c93efc8c2710207c74e79
8f0b568257b75176c56703ae57036cba1bf2cc23a5909f60d6bc9544473a553f
e4e50b35186ef0bec23c7df9473856a673dc2cf401c9b6552e540618fecc6954
a56bf13e0afae34cb58cf8065a08b52d714659cbe3f45e217024495c27146186
```

</details>

Sjors commented on src/consensus/merkle.cpp:
```diff
@@ -105,18 +105,18 @@ static void MerkleComputation(const std::vector<uint256>& leaves, uint256* proot
         uint256 h = leaves[count];
-        bool matchh = count == branchpos;
+        bool matchh = count == leaf_pos;
         count++;
```
Currently `ComputeMerklePath` is only called from `TransactionMerklePath` which sets `leaves.resize(block.vtx.size())`, so there's no way the array can be bigger.

Added a commit to check it anyway.

Sjors committed:
Check leaves size maximum in MerkleComputation

Belt and suspenders for future code changes.

Currently this function is only called from TransactionMerklePath() which sets leaves to the block transactions, so the Assume always holds.

Sjors commented:
Addressed the two comments and rebased.

tdb3 reviewed (approved):
code review re-ACK f86678156a3d3ce6488b383a2d6d91d28fcd2b73

Other than the rebase, the only differences observed were the `{` movement and the addition of the `Assume(leaves.size() <= UINT32_MAX);` in the last commit.

```
$ git range-diff f07a533dfcb172321972e5afb3b38a4bd24edb87..058862581085316927287817b2af01e8f4765a1d 39d3b538e6a2af8db85077e958970cdcd0ee7f7d~..f86678156a3d3ce6488b383a2d6d91d28fcd2b73
```

itornaza reviewed (approved):
re ACK f86678156a3d3ce6488b383a2d6d91d28fcd2b73

ryanofsky reviewed (approved):
Code review ACK f86678156a3d3ce6488b383a2d6d91d28fcd2b73 only changes since last review are a whitespace change and adding an Assume statement to check for size_t -> uint32_t overflows
